### PR TITLE
MelFilterBank CPU kernel

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

#### Why we need this PR?
- Need for a kernel that will produce a MelSpectrogram from a Spectrogram

#### What happened in this PR?
- Add implementation of MelFilterBank CPU kernel
- Add unit tests for MelFilterBank CPU kernel

**JIRA TASK**: [DALI-1175]

### CODE REVIEW

File: dali/kernels/audio/mel_scale/mel_filter_bank_args.h [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+#ifndef DALI_KERNELS_AUDIO_MEL_SCALE_MEL_FILTER_BANK_ARGS_H_
+#define DALI_KERNELS_AUDIO_MEL_SCALE_MEL_FILTER_BANK_ARGS_H_
+
+namespace dali {
+namespace kernels {
+namespace audio {
+
+enum class MelScaleType {
+  SLANEY,
```
Comment (mzient) [inline dali/kernels/audio/mel_scale/mel_filter_bank_args.h]:
```suggestion
  Slaney,
```
...to add to our discussion ;)

File: dali/kernels/audio/mel_scale/mel_filter_bank_args.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  HTK
+};
+
+struct MelFilterBankArgs {
+  // Signal's sampling rate
```
Comment (mzient) [inline dali/kernels/audio/mel_scale/mel_filter_bank_args.h]: Use `///` to generate Doxygen.

File: dali/kernels/audio/mel_scale/mel_filter_bank_args.h [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+// limitations under the License.
+
+#ifndef DALI_KERNELS_AUDIO_MEL_SCALE_MEL_FILTER_BANK_ARGS_H_
+#define DALI_KERNELS_AUDIO_MEL_SCALE_MEL_FILTER_BANK_ARGS_H_
+
+namespace dali {
+namespace kernels {
+namespace audio {
+
+enum class MelScaleType {
```
Comment (mzient) [inline dali/kernels/audio/mel_scale/mel_filter_bank_args.h]:
```suggestion
enum class MelScaleFormula {
```
?

File: dali/kernels/audio/mel_scale/mel_filter_bank_args.h [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+  // Number of bins used in the FFT to build the input spectrogram
+  // It should match the shape of the input data: `input.shape[axis] == nfft/2 + 1`
+  int nfft = -1;
+
+  // Determines the formula used in for mel_to_hz and hz_to_mel calculations
+  MelScaleType mel_formula = MelScaleType::SLANEY;
+
+  // Determines whether to normalize the filter weights by the width of the mel band
+  bool norm_filters = true;
```
Comment (mzient) [inline dali/kernels/audio/mel_scale/mel_filter_bank_args.h]:
```suggestion
  bool normalize = true;
```

File: dali/kernels/audio/mel_scale/mel_filter_bank_cpu.cc [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+template <typename T, int Dims>
+MelFilterBankCpu<T, Dims>::~MelFilterBankCpu() = default;
+
+template <typename T, int Dims>
+KernelRequirements MelFilterBankCpu<T, Dims>::Setup(
+    KernelContext &context,
+    const InTensorCPU<T, Dims> &in,
+    const MelFilterBankArgs &args) {
+  auto axis = args.axis >= 0 ? args.axis : Dims - 2;
```
Comment (mzient) [inline dali/kernels/audio/mel_scale/mel_filter_bank_cpu.cc]: Why not just int? It's not like there's any significant arithmetic going on here.

File: dali/kernels/audio/mel_scale/mel_filter_bank_args.h [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  SLANEY,
+  HTK
+};
+
+struct MelFilterBankArgs {
+  // Signal's sampling rate
+  float sample_rate = 44100.0f;
+
+  // Minimum frequency
+  float fmin = 0.0f;
```
Comment (mzient) [inline dali/kernels/audio/mel_scale/mel_filter_bank_args.h]: Nitpick: I'd use f_min and f_max - or even freq_min, freq_max. fmin looks like `float fmin(float a, float b)`

File: dali/kernels/audio/mel_scale/mel_filter_bank_cpu_impl.h [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+      hz = fmin + mel * fsp;
+    }
+    return hz;
+  }
+};
+
+template <typename T>
+class MelFilterBankImpl {
+ public:
+  virtual ~MelFilterBankImpl() = default;
```
Comment (mzient) [inline dali/kernels/audio/mel_scale/mel_filter_bank_cpu_impl.h]: If you just want to hide the implementation, then you don't need a virtual interface.

File: dali/kernels/audio/mel_scale/mel_filter_bank_cpu_impl.h [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+// Each triangular filter is defined by three points in this grid (left, center, right)
+// For each iteration we process a range between two mel frequencies in the grid, calculating
+// the contribution of each FFT bin to 2 triangular filters (one is in the negative slope region
+// and the other in the positive slope region), except for the first and last iteration.
+// In total, we do a single pass on every FFT bin column
+//
+// For every FFT bin we compute the weight for each filter and travel through the row, computing
+// the contributions on every window of the spectrogram (horizontal axis)
+//
+template <typename T, typename MelScale = HtkMelScale<T>>
```
Comment (mzient) [inline dali/kernels/audio/mel_scale/mel_filter_bank_cpu_impl.h]: Unless you envision totally replacing this implementation based on MelScale, then this argument doesn't need to be here at all - it's only needed in constructor.

File: dali/kernels/audio/mel_scale/mel_filter_bank_cpu_impl.h [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+  int nfft_;
+  T sample_rate_;
+  T freq_low_;
+  T freq_high_;
+  bool norm_filters_;
+
+  T mel_low_;
+  T mel_high_;
+  T hz_step_;
+  T mel_delta_;
```
Comment (mzient) [inline dali/kernels/audio/mel_scale/mel_filter_bank_cpu_impl.h]: Many of these fields are not used beyond the scope of the constructor.

File: dali/kernels/audio/mel_scale/mel_filter_bank_args.h (hunk 1 above)
Comment (jantonguirao) [reply dali/kernels/audio/mel_scale/mel_filter_bank_args.h]: done

File: dali/kernels/audio/mel_scale/mel_filter_bank_args.h (hunk 3 above)
Comment (jantonguirao) [reply dali/kernels/audio/mel_scale/mel_filter_bank_args.h]: done

File: dali/kernels/audio/mel_scale/mel_filter_bank_args.h (hunk 2 above)
Comment (jantonguirao) [reply dali/kernels/audio/mel_scale/mel_filter_bank_args.h]: done

File: dali/kernels/audio/mel_scale/mel_filter_bank_cpu.cc (hunk 5 above)
Comment (jantonguirao) [reply dali/kernels/audio/mel_scale/mel_filter_bank_cpu.cc]: done

File: dali/kernels/audio/mel_scale/mel_filter_bank_args.h (hunk 4 above)
Comment (jantonguirao) [reply dali/kernels/audio/mel_scale/mel_filter_bank_args.h]: done

File: dali/kernels/audio/mel_scale/mel_filter_bank_args.h (hunk 6 above)
Comment (jantonguirao) [reply dali/kernels/audio/mel_scale/mel_filter_bank_args.h]: done

File: dali/kernels/audio/mel_scale/mel_filter_bank_cpu_impl.h (hunk 7 above)
Comment (jantonguirao) [reply dali/kernels/audio/mel_scale/mel_filter_bank_cpu_impl.h]: done

File: dali/kernels/audio/mel_scale/mel_filter_bank_cpu_impl.h (hunk 8 above)
Comment (jantonguirao) [reply dali/kernels/audio/mel_scale/mel_filter_bank_cpu_impl.h]: done

File: dali/kernels/audio/mel_scale/mel_filter_bank_cpu_impl.h (hunk 9 above)
Comment (jantonguirao) [reply dali/kernels/audio/mel_scale/mel_filter_bank_cpu_impl.h]: done

File: dali/kernels/audio/mel_scale/mel_filter_bank_cpu.h [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+                                      const MelFilterBankArgs &args);
+
+  DLL_PUBLIC void Run(KernelContext &context,
+                      const OutTensorCPU<T, Dims> &out,
+                      const InTensorCPU<T, Dims> &in,
+                      const MelFilterBankArgs &args);
+
+ private:
+  class Impl;
+  std::shared_ptr<Impl> impl_;
```
Comment (mzient) [inline dali/kernels/audio/mel_scale/mel_filter_bank_cpu.h]: In this case, please delete or implement copy and assignment.

Comment (mzient) [reply dali/kernels/audio/mel_scale/mel_filter_bank_cpu.h]: Just discovered - you just need out-of-place default constructor to use unique_ptr here. Will do the same in my PR.

Comment (jantonguirao) [reply dali/kernels/audio/mel_scale/mel_filter_bank_cpu.h]: Applied, thanks!

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

**Pull request merged**